torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1117,6 +1119,153 @@ def arg_extractor(combine_fn, input, dim):
+            (combine_result, combine_treespec),
+            combine_graph,
+            combine_lifted_freevars,
+        ) = speculate_subgraph(
+            tx,
+            combine_fn,
+            sub_args,
+            sub_kwargs={},
+            description="scan_combine",
+            source_target=self.value,
+            set_subgraph_inputs="flatten_manual",
+        )
+
+        if combine_lifted_freevars:
+            unimplemented(
+                f"Combine fn had unexpected freevars: {combine_lifted_freevars}"
+            )
+
+        if combine_result.python_type() != list:
+            unimplemented(
+                f"Expected combine_fn to return a list if tensor but got {combine_result.python_type()}",
+            )
+
+        input_proxy = input.as_proxy()
+        init_proxy = init.as_proxy()
+        combine_result_proxy = combine_result.as_proxy()
+        for result, inp_proxy, ini_proxy, s_args in zip(
+            combine_result_proxy, input_proxy, init_proxy, sub_args
+        ):
+            inp_meta = inp_proxy.node.meta["example_value"]
+            ini_meta = ini_proxy.node.meta["example_value"]
+            combine_result_meta = result.node.meta["example_value"]
+            s_args = s_args.as_proxy().node.meta["example_value"]
+            if combine_result_meta.device != inp_meta.device:
+                unimplemented(
+                    f"Expected combine_fn to return a tensor on device {inp_meta.device} but "
+                    + f"got {combine_result_meta.device}"
+                )
+            if combine_result_meta.dtype != inp_meta.dtype:
+                unimplemented(
+                    f"Expected combine_fn to return a tensor of {inp_meta.dtype} but "
+                    + f"got {combine_result_meta.dtype}"
+                )
+            if combine_result_meta.shape != s_args.shape:
+                unimplemented(
+                    f"Expected shape of input with {s_args.shape} to be the same as "
+                    + f"combine_result with {combine_result_meta.shape}"
+                )
+            if (
+                combine_result_meta.device != ini_meta.device
+                or combine_result_meta.dtype != ini_meta.dtype
+                or combine_result_meta.shape != ini_meta.shape
+            ):
+                unimplemented(
+                    f"Expected metadata of the combine_fn result {combine_result_meta} to be the same as "
+                    + f"the metadata of init with {inp_meta}"
+                )
+
+        combine_gm = torch.fx.GraphModule(dict(tx.output.nn_modules), combine_graph)
+        combine_fn_name = add_subgraph(tx, "scan_combine", combine_gm)
+
+        p_args = (
+            make_attr(tx, combine_fn_name),
+            input_proxy,
+            init_proxy,
+            dim.as_proxy(),
+        )
+
+        with tx.fake_mode:
+            out_meta = tuple(
+                ini_proxy.node.meta["example_value"].clone() for ini_proxy in init_proxy
```

Comment:
This is probably wrong? The example output of the scan operator should be the the inits stacked together, right?  We have similar logic for map in https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/higher_order_ops.py#L1185-L1196

To test this, we can have a downstream operator that takes the output of scan as output e.g.
suppose scan() -> (3, 2), we call Tensor (3, 3)  @ (3, 2).

Reply:
Good catch!
The code here would still work as init necessarily has a `1` at the dim of scan and then later automatic broadcasting should take care of  properly expanding it. However, I agree with you that we should have a correct version and implemented one taking inspiration from map.
I also added a testcase, [test_pointwise_scan_downstream_op](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/test/functorch/test_control_flow.py#L1555)